File: day-19/main.py
import math
import operator

op_lambda = {
    'add': operator.add,
    'mul': operator.mul,
    'ban': operator.and_,
    'bor': operator.or_,
    'gt': operator.gt,
    'eq': operator.eq,
}


def parse(line):
    parts = line.split()
    return (parts[0],) + tuple(int(_) for _ in parts[1:])


def exec_op(op_code, register):
    op_type, a, b, c = op_code
    register = register.copy()
    if op_type[:3] in ['add', 'mul', 'ban', 'bor']:
        a = register[a]
        b = register[b] if op_type[3] == 'r' else b
        register[c] = op_lambda[op_type[:3]](a, b)
    elif op_type[:3] == 'set':
        a = register[a] if op_type[3] == 'r' else a
        register[c] = a
    elif op_type[:2] in ['gt', 'eq']:
        a = register[a] if op_type[2] == 'r' else a
        b = register[b] if op_type[3] == 'r' else b
        register[c] = int(op_lambda[op_type[:2]](a, b))
    return register


def exec_program(program, ip_reg, init_registers, breakpoint=None):
    registers = init_registers.copy()
    ip = 0
    while 0 <= ip < len(program):
        op = program[ip]
        registers[ip_reg] = ip
        registers = exec_op(op, registers)
        if breakpoint is not None and ip == breakpoint:
            return registers
        ip = registers[ip_reg] + 1
    return registers


def sum_divisors(x):
    s = 0
    for i in range(1, int(math.sqrt(x))):
        if x % i == 0:
            s += i
            if x // i != i:
                s += x // i
    return s

def main():
    with open('input.txt') as f:
        lines = f.read().strip().split('\n')
    program = [parse(line) for line in lines[1:]]
    ip_reg = int(lines[0].split()[1])

    # Part 1
    registers = exec_program(program, ip_reg, [0] * 6, breakpoint=1)
    print(sum_divisors(registers[5]))

    # Part 2
    registers = exec_program(program, ip_reg, [1] + [0] * 5, breakpoint=1)
    print(sum_divisors(registers[5]))


if __name__ == '__main__':
    main()

File: day-19/src/main.rs
use std::fs;

const N_REGS: usize = 6;

type Registers = [i32; N_REGS];

struct Instruction(String, i32, i32, i32);

fn clone<I>(iter: I) -> [i32; N_REGS] where
        I: Iterator<Item=i32> {
    let mut arr = [0; N_REGS];
    for (i, x) in iter.enumerate() {
        arr[i] = x;
    }
    return arr;
}

fn parse_i32(s: &str) -> i32 {
    s.parse::<i32>().ok().unwrap()
}

fn parse_instruction(s: &str) -> Instruction {
    let mut iter = s.split_whitespace();
    let op = iter.next().unwrap().to_string();
    let params = iter.map(parse_i32).collect::<Vec<_>>();
    Instruction(op, params[0], params[1], params[2])
}

fn exec(ins: &Instruction, regs: Registers) -> Registers {
    let mut new_regs = clone(regs.iter().map(|x| *x));
    let op_type = &ins.0;
    let (a, b, c) = (ins.1, ins.2, ins.3 as usize);
    match &op_type[..2] {
        "ad" | "mu" | "ba" | "bo" => {
            let a = regs[a as usize];
            let b = if &op_type[3..4] == "r" { regs[b as usize] } else { b };
            new_regs[c] = match &op_type[..3] {
                "add" => a + b,
                "mul" => a * b,
                "ban" => a & b,
                "bor" => a | b,
                _ => { panic!(); }
            };
        }
        "se" => {
            let a = if &op_type[3..4] == "r" { regs[a as usize] } else { a };
            new_regs[c] = a;
        }
        "gt" | "eq" => {
            let a = if &op_type[2..3] == "r" { regs[a as usize] } else { a };
            let b = if &op_type[3..4] == "r" { regs[b as usize] } else { b };
            new_regs[c] = match &op_type[..2] {
                "gt" => (a > b) as i32,
                "eq" => (a == b) as i32,
                _ => { panic!(); }
            };
        }
        _ => { panic!(); }
    };
    return new_regs;
}

fn exec_program(program: &Vec<Instruction>, ip_reg: usize,
                init_regs: Registers, breakpoint: Option<usize>) -> Registers {
    let mut regs = clone(init_regs.iter().map(|x| *x));
    let mut ip = 0i32;
    while ip >= 0 && ip < program.len() as i32 {
        regs[ip_reg] = ip;
        regs = exec(&program[ip as usize], regs);
        if breakpoint.is_some() && breakpoint.unwrap() == ip as usize {
            return regs;
        }
        ip = regs[ip_reg] + 1;
    }
    return regs;
}

fn sum_divisors(x: usize) -> usize {
    let mut sum = 0;
    for i in 1..((x as f64).sqrt() as usize) {
        if x % i != 0 { continue; }
        sum += i;
        let d = x / i;
        if d != i { sum += d; }
    }
    return sum;
}

fn main() {
    let input = fs::read_to_string("day-19/input.txt").unwrap();
    let mut lines = input.lines();
    let ip_reg = lines.next().unwrap()
            .split_whitespace().skip(1).next().unwrap()
            .parse::<usize>().unwrap();
    let program: Vec<Instruction> = lines.map(parse_instruction).collect::<Vec<_>>();

    // Part 1
    let regs = [0; N_REGS];
    let regs = exec_program(&program, ip_reg, regs, Some(1));
    println!("{}", sum_divisors(regs[5] as usize));

    // Part 2
    let mut regs = [0; N_REGS];
    regs[0] = 1;
    let regs = exec_program(&program, ip_reg, regs, Some(1));
    println!("{}", sum_divisors(regs[5] as usize));
}
